lib/mongo/server/connection_pool/queue.rb

```diff
@@ -313,19 +329,86 @@ def close_stale_sockets!
             end
           end
         ensure
-          check_count_invariants
+          check_count_invariants(false)
         end
 
         private
 
+        def close_connection!(connection, reason)
+          connection.disconnect!
+          publish_cmap_event(
+            Monitoring::Event::ConnectionClosed.new(
+              reason,
+              @address,
+              connection.id,
+            )
+          )
+
+          @pool_size = if @pool_size > 0
+            @pool_size - 1
+          else
+            # This should never happen
+            log_warn("ConnectionPool::Queue: unexpected enqueue")
+            0
+          end
+        end
+
+        def is_stale?(connection)
+          if connection.generation != @generation
+            close_connection!(connection, Monitoring::Event::ConnectionClosed::STALE)
+            true
+          end
+        end
+
+        def is_idle?(connection)
+          if connection && connection.last_checkin && max_idle_time
+            if Time.now - connection.last_checkin > max_idle_time
+              close_connection!(connection, Monitoring::Event::ConnectionClosed::IDLE)
+              true
+            end
+          end
+        end
+
         def dequeue_connection
+          semaphore = check_wait_queue
           deadline = Time.now + wait_timeout
-          loop do
-            return queue.shift unless queue.empty?
-            connection = create_connection
-            return connection if connection
-            wait_for_next!(deadline)
+          semaphore.wait(wait_timeout) if semaphore
+          raise Error::WaitQueueTimeout.new(@address) if deadline <= Time.now
+
+          mutex.synchronize do
+            loop do
+              until queue.empty?
+                connection = queue.shift
+
+                unless is_stale?(connection) || is_idle?(connection)
+                  return connection
+                end
+              end
+
+              connection = create_connection
+              return connection if connection
+
+              wait = deadline - Time.now
+              resource.wait(mutex, wait)
+              raise Error::WaitQueueTimeout.new(@address) if deadline <= Time.now
+            end
+          end
+        end
+
+        def check_wait_queue
```

Comment:
What does this method do?

Reply:
It adds a thread to the wait queue, and, if necessary, returns a semaphore to signal the thread when to wake up and acquire a connection.

Reply:
Please add a docstring containing this information, as well as when the semaphore is returned and when it is not returned.

Reply:
We don't currently document any of existing private methods in the Queue; I don't see why this one is any different, especially given that it's much shorter, so it's already obvious what's being returned.

Reply:
This method is gone now, so I'm going to resolve this